### PR TITLE
GMF print

### PR DESCRIPTION
**Pull request by ger-benjamin**
Rel to: https://github.com/camptocamp/c2cgeoportal/issues/1660
Specifications: https://github.com/camptocamp/c2cgeoportal/wiki/Spec-%231660-Print

Example: 
 - https://ger-benjamin.github.io/ngeo/gmfprint/examples/contribs/gmf/apps/desktop 
 - https://ger-benjamin.github.io/ngeo/gmfprint/examples/contribs/gmf/apps/desktop_alt

 (not working because of: `XMLHttpRequest cannot load https://geomapfish-demo.camptocamp.net/2.0/wsgi/printproxy/report.pdf. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'https://ger-benjamin.github.io' is therefore not allowed access. The response had HTTP status code 500.`)

I suggest that I resolve stalled points and issues in another PR

### Stalled tasks
 - Add the selected features to the print report :point_right: Wait on a query component. -->  --> https://github.com/camptocamp/ngeo/issues/1104

### Know issues:
 - An 'out of scale' symbol will have is legend displayed. Perhaps that this problem must be handle on server side (that already exists with cgxp).

#### Further improvements
 - ~~'Scales' field has no formatting (`100000` should be `1:100'000`). I think that the better we can do is to create a filter for this input. But if we want to creates it properly, we need a "filter" folder in ngeo/src/. That implies to adapt the doc generation and to add these filters in the compiled folder. So I want to do that in another PR.~~ --> I'll put it temporarily into `ngeo/src/service` but that must be improved. (Tests, documentations, etc.) --> https://github.com/camptocamp/ngeo/issues/1106
 - More tests for the print controller --> https://github.com/camptocamp/ngeo/issues/836
 -  https://github.com/camptocamp/ngeo/issues/1105

### CODE REVIEW

**Comment by ger-benjamin** [comment]
Can be reviewed even if it is not completely finished.

**Comment by sbrunner** [comment]
The rotation with Alt+Shift don't works ...

**Comment by sbrunner** [comment]
> not working because of: XMLHttpRequest cannot load https://geomapfish-demo.camptocamp.net/2.0/wsgi/printproxy/report.pdf. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'https://ger-benjamin.github.io' is therefore not allowed access. The response had HTTP status code 500.

Fixed :-)

**Comment by sbrunner** [comment]
It should use the right DPI...

**Comment by sbrunner** [comment]
CTRL+SHIFT+Drag => no more print extent ...

**Comment by sbrunner** [comment]
Wrong layer name in the Legend (should be translated...)

**Comment by ger-benjamin** [comment]
Ready for a review

File: contribs/gmf/less/print.less [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    cursor: pointer;
+    height: 5px;
+  }
+  input[type=range]:focus::-ms-fill-lower {
+    background: rgba(148, 172, 154, 0.78);
+  }
+  input[type=range]:focus::-ms-fill-upper {
+    background: rgba(205, 217, 208, 0.78);
+  }
+}
```
Comment (pgiraud) [inline contribs/gmf/less/print.less]: The input range styling should be made available not only for the print panel in my opinion but for all the application.

File: contribs/gmf/src/directives/partials/print.html [hunk 2]
```diff
@@ -0,0 +1,69 @@
+<div class="print-panel">
+  <div ng-show="ctrl.isState('CAPABILITIES_NOT_LOADED') !== true && ctrl.isState('ERROR_ON_GETCAPABILITIES') !== true">
+    <div ng-repeat="customField in ctrl.fields.customs" class="form-group">
+      <input ng-if="customField.type === 'text'" ng-model="customField.value" class="form-control" type="text" placeholder="{{customField.name|translate}}">
+      <textarea ng-if="customField.type === 'textarea'" ng-model="customField.value" class="form-control" row="3" placeholder="{{customField.name|translate}}"></textarea>
+      <p ng-if="customField.type === 'checkbox'"><input ng-model="customField.value" type="checkbox"> <span translate>Debug</span></p>
```
Comment (pgiraud) [inline contribs/gmf/src/directives/partials/print.html]: What is this "Debug" checkbox for?

File: contribs/gmf/src/directives/partials/print.html [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      <label>
+        <input ng-model="ctrl.fields.legend" type="checkbox">
+        <span translate>Legend</span>
+      </label>
+    </div>
+    <div class="form-horizontal">
+      <div ng-show="ctrl.fields.layouts.length !== 1" class="form-group form-group-sm">
+        <label class="control-label col-md-3" translate>Layout</label>
+        <div class="col-md-9">
+            <button type="button" class="btn btn-default form-control" data-toggle="dropdown">{{ctrl.fields.layout | translate}}</button>
```
Comment (pgiraud) [inline contribs/gmf/src/directives/partials/print.html]: indentation

File: contribs/gmf/src/directives/partials/print.html [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+          <ul class="dropdown-menu" role="menu">
+              <li ng-repeat="layout in ctrl.fields.layouts"><a href="" ng-click="ctrl.setLayout(layout)">{{layout | translate}}</a></li>
+          </ul>
+        </div>
+      </div>
+      <div ng-show="ctrl.fields.scales.length !== 1" class="form-group form-group-sm">
+        <label class="control-label col-md-3"translate>Scale</label>
+        <div class="col-md-9">
+            <button type="button" class="btn btn-default form-control" data-toggle="dropdown" aria-expanded="false">{{ctrl.fields.scale | translate}}</button>
```
Comment (pgiraud) [inline contribs/gmf/src/directives/partials/print.html]: indentation

File: contribs/gmf/src/directives/partials/print.html [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        <span translate>Legend</span>
+      </label>
+    </div>
+    <div class="form-horizontal">
+      <div ng-show="ctrl.fields.layouts.length !== 1" class="form-group form-group-sm">
+        <label class="control-label col-md-3" translate>Layout</label>
+        <div class="col-md-9">
+            <button type="button" class="btn btn-default form-control" data-toggle="dropdown">{{ctrl.fields.layout | translate}}</button>
+          <ul class="dropdown-menu" role="menu">
+              <li ng-repeat="layout in ctrl.fields.layouts"><a href="" ng-click="ctrl.setLayout(layout)">{{layout | translate}}</a></li>
```
Comment (pgiraud) [inline contribs/gmf/src/directives/partials/print.html]: indentation

File: contribs/gmf/src/directives/partials/print.html (hunk 3 above)
Comment (pgiraud) [inline contribs/gmf/src/directives/partials/print.html]: This button is missing a caret.

File: contribs/gmf/src/directives/partials/print.html [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            <button type="button" class="btn btn-default form-control" data-toggle="dropdown" aria-expanded="false">{{ctrl.fields.scale | translate}}</button>
+          <ul class="dropdown-menu" role="menu">
+            <li ng-repeat="scale in ctrl.fields.scales"><a href="" ng-click="ctrl.setScale(scale)">{{scale | translate}}</a></li>
+          </ul>
+        </div>
+      </div>
+      <div ng-show="ctrl.fields.dpis.length !== 1" class="form-group form-group-sm">
+        <label class="control-label col-md-3" translate>DPI</label>
+        <div class="col-md-5">
+          <button type="button" class="btn btn-default form-control" data-toggle="dropdown" aria-expanded="false">{{ctrl.fields.dpi}}</button>
```
Comment (pgiraud) [inline contribs/gmf/src/directives/partials/print.html]: This button is missing a caret.

File: contribs/gmf/src/directives/partials/print.html [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+          </ul>
+        </div>
+      </div>
+      <div class="form-group form-group-sm">
+        <label class="control-label col-md-3" translate>Rotation</label>
+        <div class="col-md-5">
+          <input class="form-control" type="range" ng-model="ctrl.getSetRotation" ng-model-options="{getterSetter: true}" min="-180" max="180" style="border:0; background-color: transparent; box-shadow: none;" data-toggle="tooltip" title="{{'You can also use Alt+Shift on the map' | translate}}">
+        </div>
+        <div class="col-md-4">
+          <input class="form-control" type="number" ng-model="ctrl.getSetRotation" ng-model-options="{getterSetter: true}">
```
Comment (pgiraud) [inline contribs/gmf/src/directives/partials/print.html]: The number may not be correctly displayed.
screenshot from 2016-04-27 13 20 11

File: contribs/gmf/src/directives/partials/print.html [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        </div>
+      </div>
+    </div>
+    <div class="form-group pull-right">
+      <div class="col-md-1" ng-show="ctrl.isState('PRINTING')"><i class="fa fa-refresh fa-spin"></i></div>
```
Comment (pgiraud) [inline contribs/gmf/src/directives/partials/print.html]: Spinner is not correctly vertically aligned.
screenshot from 2016-04-27 13 23 51

File: contribs/gmf/externs/gmfx.js [hunk 9]
```diff
@@ -43,6 +43,119 @@ gmfx.PermalinkOptions.prototype.crosshairStyle;
...
+
+/**
+ * The list of 'layouts'.
+ * @type {Array.<string>|undefined}
+ */
+gmfx.PrintFields.prototype.layouts;
+
+
+/**
+ * The list of 'scales'.
```
Comment (fredj) [inline contribs/gmf/externs/gmfx.js]: legend instead

File: contribs/gmf/externs/gmfx.js [hunk 10]
```diff
@@ -43,6 +43,119 @@ gmfx.PermalinkOptions.prototype.crosshairStyle;
...
+
+/**
+ * The selected 'scale'.
+ * @type {number|undefined}
+ */
+gmfx.PrintFields.prototype.scale;
+
+
+/**
+ * The 'debug' field.
```
Comment (fredj) [inline contribs/gmf/externs/gmfx.js]: `The list of 'scales'.` instead

File: contribs/gmf/src/directives/print.js [hunk 11]
```diff
@@ -0,0 +1,803 @@
...
+
+    if (layer.getVisible() && source !== undefined) {
+      // For WMTS layers.
+      if (layer instanceof ol.layer.Tile) {
+        layerName = layer.get('layerName');
+        icons = this.ngeoLayerHelper_.getWMTSLegendURL(layer);
+        // Don't add classes without legend url.
+        if (icons !== null) {
+          classes.push({
+            'name': gettextCatalog_.getString('' + layerName),
```
Comment (fredj) [inline contribs/gmf/src/directives/print.js]: I think that the variable should be `gettextCatalog` to work (only `gettextCatalog.getString` is recognized)

Comment (fredj) [inline contribs/gmf/src/directives/print.js]: Forget last comment; it's for the string collect only

File: src/services/print.js [hunk 12]
```diff
@@ -182,20 +195,21 @@ ngeo.Print.prototype.encodeMap_ = function(map, scale, object) {
   var viewCenter = view.getCenter();
   var viewProjection = view.getProjection();
   var viewResolution = view.getResolution();
-  var viewRotation = view.getRotation();
+  var viewRotation = object.rotation || view.getRotation() * 180 / Math.PI;
```
Comment (fredj) [inline src/services/print.js]: minor but `ol.math.toDegrees(view.getRotation())` can be used here

File: contribs/gmf/src/directives/partials/print.html (hunk 2 above)
Comment (ger-benjamin) [inline contribs/gmf/src/directives/partials/print.html]: Oups !
That's a custom field. I must display the name of the field. Not "Debug".

File: contribs/gmf/src/directives/partials/print.html (hunk 3 above)
Comment (pgiraud) [inline contribs/gmf/src/directives/partials/print.html]: Also the value should be "1 : xxxx". It's a scale not only a denominator.

Comment (sbrunner) [inline contribs/gmf/src/directives/partials/print.html]: with a coma only for english people ...

File: contribs/gmf/src/directives/partials/print.html (hunk 7 above)
Comment (ger-benjamin) [inline contribs/gmf/src/directives/partials/print.html]: Done but ~badly~ I've no clue on how to do that properly with all the contraints. So I've change the total size of the right panel from 25 rem to 28 rem. The gut point is that the width of the new right panel (28rem) + the width of the "tools" right panel (the little one, at the right) (4rem) = 32 rem. And the left panel width is 32 rem too.

**Comment by fredj** [comment]
Looks good, what about merging and improve (if needed) the numbers input later? @pgiraud ?

**Comment by ger-benjamin** [comment]
(I'm currently adding tests for exported gmf print controller functions. But +1 with @fredj . This PR is huge and I'm scared at each rebase).

File: contribs/gmf/examples/print.html [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+          gmf-layertree-openlinksinnewwindow="true">
+      </gmf-layertree>
+    </div>
+
+    <div class="clear-left"></div>
+
+    <p id="desc">
+      This example shows how to use the <code>gmf.printDirective</code>.
+      Directive that generates fields and manages tools to print with a
+      mapserver print v3.
```
Comment (tsauerwein) [inline contribs/gmf/examples/print.html]: MapFish Print 3

File: contribs/gmf/src/directives/print.js [hunk 14]
```diff
@@ -0,0 +1,803 @@
...
+      this.map.on('pointerdrag', this.onPointerDrag_.bind(this));
+      this.map.render();
+    }.bind(this), function(resp) {
+      // Get capabilities - On error
+      this.printState = gmf.PrintState.ERROR_ON_GETCAPABILITIES;
+    }.bind(this));
+
+  } else {
+    this.map.un('postcompose', this.postcomposeListener_);
+    this.map.un('pointerdrag', this.onPointerDrag_.bind(this));
```
Comment (tsauerwein) [inline contribs/gmf/src/directives/print.js]: Does this actually work? Calling `this.onPointerDrag_.bind(this)` generates a new function every time, so the `un` does not know what function you want to unregister. Better work with listener keys.

File: contribs/gmf/src/directives/print.js [hunk 15]
```diff
@@ -0,0 +1,803 @@
...
+
+
+/**
+ * Getter setter to update or get the current rotation value. Param and result
+ *     are in degree.
+ * @param {number=} opt_rotation The optional new rotation value.
+ * @return {number} The new value of rotation;
+ * @export
+ */
+gmf.PrintController.prototype.getSetRotation = function(opt_rotation) {
```
Comment (tsauerwein) [inline contribs/gmf/src/directives/print.js]: `getSet` in one method? Why not make two methods `get` and `set`?

File: contribs/gmf/src/directives/print.js [hunk 16]
```diff
@@ -0,0 +1,803 @@
...
+  if (this.active && originalEvent.altKey && originalEvent.shiftKey) {
+    var center = this.map.getPixelFromCoordinate(this.map.getView().getCenter());
+    var pixel = e.pixel;
+    // Reset previous position between two differents sessions of drags events.
+    if (this.rotationTimeoutPromise_ === null) {
+      this.onDragPreviousMousePosition_ = null;
+    } else {
+      // Cancel the timeout to keep this session of drags event
+      this.$timeout_.cancel(this.rotationTimeoutPromise_);
+      // Calculate angle and sense of rotation.
```
Comment (tsauerwein) [inline contribs/gmf/src/directives/print.js]: It would be nice to have some tests for this logic.

File: contribs/gmf/src/directives/print.js [hunk 17]
```diff
@@ -0,0 +1,803 @@
...
+gmf.PrintController.prototype.handleGetStatusSuccess_ = function(ref, resp) {
+  var mfResp = /** @type {MapFishPrintStatusResponse} */ (resp.data);
+  var done = mfResp.done;
+  if (done) {
+    // The report is ready. Open it by changing the window location.
+    window.location.href = this.ngeoPrint_.getReportUrl(ref);
+    this.resetPrintStates_();
+  } else {
+    // The report is not ready yet. Check again in 1s.
+    var that = this;
```
Comment (tsauerwein) [inline contribs/gmf/src/directives/print.js]: Why not use `bind`?

File: contribs/gmf/src/directives/print.js [hunk 18]
```diff
@@ -0,0 +1,803 @@
...
+
+/**
+ * Set the current layout and update all fields with this new layout parameters.
+ * @param {string!} layoutName A layout name as existing in the list of
+ *     existing layouts.
+ * @export
+ */
+gmf.PrintController.prototype.setLayout = function(layoutName) {
+  var layout;
+  this.layouts_.forEach(function(l) {
```
Comment (tsauerwein) [inline contribs/gmf/src/directives/print.js]: You could use [Array.find](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) instead.

Comment (tsauerwein) [inline contribs/gmf/src/directives/print.js]: Ah no, no support for IE.

File: contribs/gmf/src/directives/print.js (hunk 15 above)
Comment (ger-benjamin) [inline contribs/gmf/src/directives/print.js]: Please search "gettersetter" on this page: https://docs.angularjs.org/api/ng/directive/ngModel

Comment (tsauerwein) [inline contribs/gmf/src/directives/print.js]: Thanks for the info.

File: contribs/gmf/src/directives/print.js (hunk 16 above)
Comment (ger-benjamin) [inline contribs/gmf/src/directives/print.js]: Yes, I'll do more tests later

**Comment by ger-benjamin** [comment]
Thanks to all. First comment updated, demo updated, code rebased.

**Comment by kalbermattenm** [comment]
Just for the record... We had a discussion with @pgiraud and @ybolognini about the print rotation and regarding the two possible ways to do this: like geo.admin.ch and lux (rotating the map) or like currently CGXP (rotating the print mask)...

@ybolognini, do you want me to discuss this matter with the PSC ? (@pgiraud, I know what you would prefer and I did get the "stiff neck" thing)

**Comment by pgiraud** [comment]
I just tested with one of my colleagues, asking him to open the url below and print the map by making sure that the longest dimension of the lake is oriented bottom to top. He inevitably turned the head. ;-)
https://ger-benjamin.github.io/ngeo/gmfprint/examples/contribs/gmf/apps/desktop/?lang=en&baselayer_ref=map&map_x=572856&map_y=197593&map_zoom=7&tree_enable_osm_open=false&tree_enable_Layer%20with%20very%20very%20very%20very%20very%20long%20name=false&tree_enable_osm_time=false&tree_enable_osm_scale=false&tree_group_layers_Layers&tree_group_layers_Group
Here's what the interface look like before printing and the result pdf.

screenshot from 2016-04-29 15-44-11

Given this, here are the arguments I have against the rotation of the frame instead of the rotation of map below:
 - user may need to lean the head to see what will be printed to have a clear view of what will be printed,
 - after some rotation it's difficult to know where the top of the square is,
 - the square may be truncated, ie. corners may be outside of the visible view, especially at 45°.

**Comment by ger-benjamin** [comment]
That's not complicated to add an option to let the integrator choose between the rotation of the map or of the mask. (Both logic are already done. Here or in ol3).

But **that's not the place to speak about that !** Please, open a thread of discussion elsewhere (like here: https://github.com/camptocamp/c2cgeoportal/issues/1660).

**Comment by ger-benjamin** [comment]
+ print rotation option.
Examples updated (example and desktop: map rotation; desktop_alt: mask rotation).

**Comment by fredj** [comment]
LGTM, the remaining enhancements can be done later (open an issue)

ping @pgiraud , @sbrunner, @tsauerwein

**Comment by sbrunner** [comment]
OK for me :-)

**Pull request merged**